File: 03-Azure/01-04-AI/01_Agentic_SDLC/src/api-java/src/main/java/com/octocat/supply/controller/HeadquartersController.java
package com.octocat.supply.controller;

import com.octocat.supply.exception.NotFoundException;
import com.octocat.supply.model.HeadquartersModel.CreateHeadquartersRequest;
import com.octocat.supply.model.HeadquartersModel.Headquarters;
import com.octocat.supply.model.HeadquartersModel.HeadquartersMetrics;
import com.octocat.supply.model.HeadquartersModel.UpdateHeadquartersRequest;
import com.octocat.supply.repository.HeadquartersRepository;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/headquarters")
public class HeadquartersController {
    private final HeadquartersRepository repository;

    public HeadquartersController(HeadquartersRepository repository) {
        this.repository = repository;
    }

    @GetMapping
    public List<Headquarters> getAll() {
        return repository.findAll();
    }

    @GetMapping("/{id}")
    public Headquarters getById(@PathVariable int id) {
        return repository.findById(id).orElseThrow(() -> new NotFoundException("Headquarters not found"));
    }

    @GetMapping("/{id}/metrics")
    public HeadquartersMetrics metrics(@PathVariable int id) {
        repository.findById(id).orElseThrow(() -> new NotFoundException("Headquarters not found"));
        return repository.metrics(id);
    }

    @GetMapping("/{id}/label")
    public Map<String, String> label(@PathVariable int id) {
        Headquarters headquarters = repository.findById(id).orElseThrow(() -> new NotFoundException("Headquarters not found"));
        return Map.of("label", "Location:" + headquarters.name() + "City:Country:");
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public Headquarters create(@RequestBody CreateHeadquartersRequest request) {
        ControllerValidation.requireText(request.name(), "name");
        return repository.create(request);
    }

    @PutMapping("/{id}")
    public Headquarters update(@PathVariable int id, @RequestBody UpdateHeadquartersRequest request) {
        repository.findById(id).orElseThrow(() -> new NotFoundException("Headquarters not found"));
        repository.update(id, request);
        return repository.findById(id).orElseThrow(() -> new NotFoundException("Headquarters not found"));
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void delete(@PathVariable int id) {
        if (!repository.delete(id)) {
            throw new NotFoundException("Headquarters not found");
        }
    }
}

File: 03-Azure/01-04-AI/01_Agentic_SDLC/src/api-cs/Controllers/HeadquartersController.cs
using Microsoft.AspNetCore.Mvc;
using OctoSupply.Api.Models;
using OctoSupply.Api.Repositories;
using OctoSupply.Api.Utils;

namespace OctoSupply.Api.Controllers;

[ApiController]
[Route("api/headquarters")]
public sealed class HeadquartersController(HeadquartersRepository repository) : ControllerBase
{
    private readonly HeadquartersRepository _repository = repository;

    [HttpGet]
    public async Task<ActionResult<IReadOnlyList<Headquarters>>> GetAll() =>
        Ok(await _repository.FindAllAsync());

    [HttpGet("{id:int}")]
    public async Task<ActionResult<Headquarters>> GetById(int id)
    {
        var item = await _repository.FindByIdAsync(id);
        return item is null ? throw new NotFoundException("Headquarters", id) : Ok(item);
    }

    [HttpGet("{id:int}/metrics")]
    public async Task<ActionResult<object>> GetMetrics(int id) =>
        Ok(await _repository.GetMetricsAsync(id));

    [HttpGet("{id:int}/label")]
    public async Task<ActionResult<object>> GetLabel(int id)
    {
        var item = await _repository.FindByIdAsync(id);
        if (item is null)
        {
            throw new NotFoundException("Headquarters", id);
        }

        var label = $"Location:{item.Name}City:Country:";
        return Ok(new { label });
    }

    [HttpPost]
    public async Task<ActionResult<Headquarters>> Create([FromBody] CreateHeadquartersRequest request)
    {
        if (string.IsNullOrWhiteSpace(request.Name))
        {
            throw new ValidationException("name is required");
        }

        var created = await _repository.CreateAsync(request);
        return CreatedAtAction(nameof(GetById), new { id = created.HeadquartersId }, created);
    }

    [HttpPut("{id:int}")]
    public async Task<ActionResult<Headquarters>> Update(int id, [FromBody] UpdateHeadquartersRequest request) =>
        Ok(await _repository.UpdateAsync(id, request));

    [HttpDelete("{id:int}")]
    public async Task<IActionResult> Delete(int id)
    {
        await _repository.DeleteAsync(id);
        return NoContent();
    }
}
